### PR TITLE
Block fetch bug fixes, QC tests and improved tracing.

### PR DESCRIPTION
**Pull request by dcoutts**
Long explanations in the individual commit messages.

Should be pretty safe to merge since it's now finally adds reasonable QC coverage and fixes bugs uncovered by those tests (some were also fixed in the previous PR, but due to these tests).

### CODE REVIEW

**Comment by mrBliss** [comment]
Since it is basically a one-line change I'm not submitting a separate PR: to fix the consensus code, rebase on master first and then apply the following change in  `Ouroboros.Consensus.Node.nodeKernel`:

```diff
nodeKernel params@NodeParams { threadRegistry } = do
    st <- initInternalState params

    forkBlockProduction st

    let IS { blockFetchInterface, fetchClientRegistry, chainDB } = st

    -- Run the block fetch logic in the background. This will call
    -- 'addFetchedBlock' whenever a new block is downloaded.
    void $ forkLinked threadRegistry $ blockFetchLogic
        nullTracer
+       nullTracer
        blockFetchInterface
        fetchClientRegistry

```

File: ouroboros-network/src/Ouroboros/Network/Testing/ConcreteBlock.hs [hunk 1]
```diff
@@ -272,6 +276,38 @@ fixupBlockHeader pb b = b'
       headerBodyHash = headerBodyHash b
     }
 
+-- |
+-- Like @'fixupBlock'@ but it will preserve the first block instead of using
+-- @GenesisHash@ and @BlockNo 1@.
+--
+fixupBlockHeader' :: (HasHeader block, HeaderHash block ~ HeaderHash Block)
+                 => Maybe block -> BlockHeader -> BlockHeader
+fixupBlockHeader' pb b = b'
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/Testing/ConcreteBlock.hs]: Would this not also work for chains that start from genesis? Perhaps we don't need two variants and this variant covers both?

File: ouroboros-network/test/Test/ChainFragment.hs [hunk 2]
```diff
@@ -372,6 +373,9 @@ prop_toChain_fromChain :: TestBlockChain -> Property
 prop_toChain_fromChain (TestBlockChain c) =
     CF.toChain (CF.fromChain c) === c
 
+prop_fixupBlock' :: TestBlockChainFragment -> Bool
+prop_fixupBlock' (TestBlockChainFragment chain) =
+  fixupChainFragment fixupBlock' (CF.toNewestFirst chain) == chain
```
Comment (dcoutts) [inline ouroboros-network/test/Test/ChainFragment.hs]: Doesn't this demonstrate that `fixupBlock'` already does the right thing for chains that start from genesis? So then we don't need two variants?

**Comment by dcoutts** [comment]
> will it make sense to use quickcheck-state-machine for testing the block fetch logic?

**Pull request merged**